modules/picard_qc/1.0/picard_qc.smk

```diff
@@ -0,0 +1,326 @@
+
+
+rule _picard_qc_merge_metrics:
+    input: 
+        _get_sample_metrics(metrics_dir = CFG["dirs"]["metrics"])
+    output: 
+        metrics = CFG["dirs"]["merged_metrics"] + "{seq_type}--{genome_build}/all.{metrics}.txt"
+    run:
+        samples = input
+        with open(output[0], "w") as out:
+            for i in range(0,len(samples)):
+                s_id = samples[i].split("/")[-2]
+                with open(samples[i], "r") as f:
+                    data = [l for l in f.readlines() if not (l.startswith('#') or l == '\n')]
+                    if i == 0:
+                        header = "SAMPLEID\t" + "\t".join(data[0].split("\t")[0:])
+                        out.write(header)
+                    if wildcards.metrics == "alignment_summary_metrics":
+                        line = s_id + "\t" + "\t".join(data[3].split("\t")[0:])
+                    else:
+                        line = s_id + "\t" + "\t".join(data[1].split("\t")[0:])
+                    out.write(line)
+                    f.close()
+        out.close()
+
+
+rule _picard_qc_flagstats:
+    input:
+        bam = rules._picard_qc_input_bam.output.sample_bam
+    output:
+        flagstats = CFG["dirs"]["metrics"] + "{seq_type}--{genome_build}/{sample_id}/flagstats"
+    log:
+        stderr = CFG["logs"]["metrics"] + "{seq_type}--{genome_build}/{sample_id}/flagstats.stderr.log"
+    conda:
+        CFG["conda_envs"]["samtools"]
+    threads:
+        CFG["threads"]["flagstats"]
+    resources:
+        mem_mb = CFG["mem_mb"]["flagstats"]
+    shell:
+        op.as_one_line("""
+        samtools flagstat -@ {threads}
+        {input.bam} > {output.flagstats} 
+        2> {log.stderr}
+        """)
+
+
+rule _picard_qc_merged_output:
+    input:
+        metrics = rules._picard_qc_merge_metrics.output.metrics
+    output:
+        metrics = CFG["dirs"]["outputs"] + "merged_metrics/{seq_type}--{genome_build}/all.{metrics}.txt"
+    run:
+        op.relative_symlink(input.metrics, output.metrics)
+
+
+rule _picard_qc_flagstats_output:
+    input:
+        flagstats = rules._picard_qc_flagstats.output.flagstats
+    output:
+        flagstats = CFG["dirs"]["outputs"] + "flagstats/{seq_type}--{genome_build}/{sample_id}.flagstats"
+    run:
+        op.relative_symlink(input.flagstats, output.flagstats)
+
+
+def _get_picard_qc_files(wildcards):
+    # base metrics to calculate for all seq_type
+    base = ["alignment_summary_metrics", "insert_size_metrics"]
+    
+    # get seq_type specific metrics
+    if wildcards.seq_type == 'mrna':
+        m = base + ["rnaseq_metrics"]
+    elif wildcards.seq_type == 'genome':
+        m = base + ["wgs_metrics"]
+    elif wildcards.seq_type == 'capture':
+        m = base + ["hs_metrics", "interval_hs_metrics"]
+    else:
+        m = base
+
+    targets = expand(rules._picard_qc_merged_output.output.metrics, **wildcards, metrics = m)
+
+    return targets
+
+ 
+rule _picard_qc_merged_dispatch:
+    input:
```

Comment:
Can you use a function call in an expand statement? If so could you call _get_picard_qc_files in the "_picard_qc_all" rule below and avoid having this extra rule?

Reply:
I think the issue is because the metrics called are different for each `seq_type` it would need to use the wildcard values. Since the target rule `picard_qc_all` doesn't have an output, it doesn't have a wildcard. There's probably a way to do it using the samples dataframe but this seems more transparent